climada/hazard/tc_surge_geoclaw.py

```diff
@@ -0,0 +1,1451 @@
+terms of the GNU Lesser General Public License as published by the Free
+Software Foundation, version 3.
+
+CLIMADA is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with CLIMADA. If not, see <https://www.gnu.org/licenses/>.
+
+---
+
+Inundation from TC storm surges, modeled using the library GeoClaw
+
+This module requires a Fortran compiler (such as gfortran) to run!
+"""
+
+import __main__
+import contextlib
+import datetime as dt
+import importlib
+import logging
+import re
+import pathlib
+import site
+import subprocess
+import sys
+import warnings
+
+import cartopy.crs as ccrs
+import cartopy.feature as cfeature
+from cartopy.mpl.gridliner import LONGITUDE_FORMATTER, LATITUDE_FORMATTER
+from matplotlib.backends.backend_agg import FigureCanvasAgg
+import matplotlib.colors
+import matplotlib.cm
+import matplotlib.pyplot as plt
+from mpl_toolkits.axes_grid1.inset_locator import inset_axes
+import numpy as np
+import pandas as pd
+import scipy.sparse as sp
+import xarray as xr
+
+from climada import CONFIG
+from climada.hazard import Centroids, Hazard, Tag as TagHazard, TropCyclone
+from climada.hazard.tc_tracks import estimate_rmw, estimate_roci
+from climada.util import ureg
+import climada.util.coordinates as u_coord
+
+LOGGER = logging.getLogger(__name__)
+
+HAZ_TYPE = 'TCSurgeGeoClaw'
+"""Hazard type acronym for this module."""
+
+CLAWPACK_GIT_URL = CONFIG.hazard.tc_surge_geoclaw.resources.clawpack_git.str()
+"""URL of the official Clawpack git repository."""
+
+CLAWPACK_VERSION = CONFIG.hazard.tc_surge_geoclaw.resources.clawpack_version.str()
+"""Version or git decorator (tag, branch) of Clawpack to use."""
+
+CLAWPACK_SRC_DIR = CONFIG.hazard.tc_surge_geoclaw.clawpack_src_dir.dir()
+"""Directory for Clawpack source checkouts (if it doesn't exist)"""
+
+GEOCLAW_WORK_DIR = CONFIG.hazard.tc_surge_geoclaw.geoclaw_work_dir.dir()
+"""Base directory for GeoClaw run data."""
+
+KN_TO_MS = (1.0 * ureg.knot).to(ureg.meter / ureg.second).magnitude
+NM_TO_KM = (1.0 * ureg.nautical_mile).to(ureg.kilometer).magnitude
+MBAR_TO_PA = (1.0 * ureg.mbar).to(ureg.pascal).magnitude
+KM_TO_DEG = 1.0 / (60 * NM_TO_KM)
+"""Unit conversion factors."""
+
+
+class TCSurgeGeoClaw(Hazard):
+    """TC storm surge heights in m, modeled using GeoClaw.
+
+    Attributes
+    ----------
+    gauge_data : list of lists of dicts
+        For each storm and each gauge, a dict containing the `location` of the gauge, and
+        (for each landfall event) `base_sea_level`, `topo_height`, `time` and `height_above_geoid`
+        information.
+        Due to this format, this data will NOT be stored when using `write_hdf5`. However, you
+        can manually pickle it in a separate file if needed.
```

Comment:
Could you maybe write a helper method `save_gauge_data` and `load_gauge_data` if this makes sense?

Reply:
Done.